### PR TITLE
Fix URL escaping for lock artifacts.

### LINKED ISSUE #2156: Pex stripping querystring from artifact downloads
Pex strips querystring from artifact URL [code](https://github.com/pantsbuild/pex/blob/adf7e5449562fc819a3e93c0f4f4d6e155284af0/pex/resolve/resolved_requirement.py#L84-L86)

We are using [pypicloud](https://pypicloud.readthedocs.io/en/latest/) for a private python package index.
This is backed by S3 and returns package links as s3 pre-signed urls, which contain querystrings for authentication with s3.

Hence the artifact download fails, when the querystrings are stripped (403 error).

Would it be possible not to strip the querystring?
The querystring changes each time, so would be best for caching if the url without the querystring is the one used for the cache.

### LINKED ISSUE #2348: URL quoting regression when using `find-links`
Matching this Pants issue; https://github.com/pantsbuild/pants/issues/20474. Users both on Slack and GitHub are reporting regressions around URL-parts not being quoted as expected when Pex tries to download artifacts. @cburroughs helpfully bisected Pex directly and has isolated the issue to commit https://github.com/pantsbuild/pex/commit/45eea4bb274b65bc185e1ea8cfc67f507209aea0.

On my own, I've confirmed using pipx that 2.1.143 works fine, and 2.1.144 is broken which matches their conclusion:
```
pex3_2_1_144 lock create --interpreter-constraint $'CPython==3.10.*' --style=universal --pip-version=23.1.2 --resolver-version pip-2020-resolver --find-links=https://download.pytorch.org/whl/torch_stable.html  "torch==2.0.1+cpu" -o tmp/tmp.lock
pid 32047 -> /home/ts/.pex/venvs/41538e837189c28a1bce8019b533d7189996d251/5985ed09b49a653d6596b0e14d134c5456cf1a9f/bin/python -sE /home/ts/.pex/venvs/41538e837189c28a1bce8019b533d7189996d251/5985ed09b49a653d6596b0e14d134c5456cf1a9f/pex --disable-pip-version-check --no-python-version-warning --exists-action a --no-input --isolated -q --cache-dir /home/ts/.pex/pip/23.1.2/pip_cache --log /tmp/pex-pip-log.fkdel996/pip.log download --dest /home/ts/.pex/downloads/fingerprint_artifact.zv_s1ok7 --no-deps https://download.pytorch.org/whl/cpu/torch-2.0.1+cpu-cp310-cp310-win_amd64.whl --index-url https://pypi.org/simple --find-links https://download.pytorch.org/whl/torch_stable.html --retries 5 --timeout 15 exited with 1 and STDERR:
  ERROR: HTTP error 403 while getting https://download.pytorch.org/whl/cpu/torch-2.0.1+cpu-cp310-cp310-win_amd64.whl
ERROR: Could not install requirement torch==2.0.1+cpu from https://download.pytorch.org/whl/cpu/torch-2.0.1+cpu-cp310-cp310-win_amd64.whl because of HTTP error 403 Client Error: Forbidden for url: https://download.pytorch.org/whl/cpu/torch-2.0.1+cpu-cp310-cp310-win_amd64.whl for URL https://download.pytorch.org/whl/cpu/torch-2.0.1+cpu-cp310-cp310-win_amd64.whl
```
There were a few unquote changes in the linked commit, but I'm not sure if those are erroneous, if we should store the original URL as well, or if we should quote them somewhere later.

### PR DESCRIPTION
**Pull request by jsirois**
Fixes #2348
Fixes #2156

### CODE REVIEW

**Comment by jsirois** [comment]
Sorry for the noise. There is some sloppy here I'd like to clean up. I'll convert to draft and ping you tomorrow when it's solid.

**Comment by jsirois** [comment]
Alright, this is ready for review now. The 1st two commits fix CI bitrot and are off for review in #2351. This review is just of [cd4db97](https://github.com/pantsbuild/pex/pull/2349/commits/cd4db979fac9d90a3aa2ae1c6cd0755b284fb25d) (the meat), and [b70b6ab](https://github.com/pantsbuild/pex/pull/2349/commits/b70b6ab759fbf014ccfc6d9f1d3095614a9dad24) (which fell out cleanly).

File: pex/resolve/resolved_requirement.py [hunk 1]
```diff
@@ -77,29 +78,72 @@ def from_hashing_fingerprint(cls, fingerprint):
...
+            path=path,
+            fragment_parameters=fragment_parameters,
+            fingerprints=tuple(fingerprints),
         )
 
     raw_url = attr.ib(eq=False)  # type: str
+    download_url = attr.ib(eq=False)  # type: str
     normalized_url = attr.ib()  # type: str
-    scheme = attr.ib()  # type: Optional[Union[str, ArchiveScheme.Value, VCSScheme]]
+    scheme = attr.ib(eq=False)  # type: Optional[Union[str, ArchiveScheme.Value, VCSScheme]]
```
Comment (benjyw) [inline pex/resolve/resolved_requirement.py:142]: Why is this necessary?

**Review by benjyw — APPROVED** [review]
Quite the blast radius. Hopefully the type annotations helped find it.

File: pex/resolve/resolved_requirement.py (hunk 1 above)
Comment (jsirois) [reply pex/resolve/resolved_requirement.py:142]: It wasn't, but neither was its redundant role in equality as part of the normalized URL either.
